runtime/ft_test.go

```diff
@@ -766,3 +770,408 @@ func BenchmarkRuntimeFungibleTokenTransfer(b *testing.B) {
+   }
+
+   init() {
+       self.totalSupply = 0.0
+   }
+}
+`
+
+func importsAddressLocation(program *ast.Program, address common.Address, name string) bool {
+	importDeclarations := program.ImportDeclarations()
+
+	// Check if the location is imported by any import declaration
+	for _, importDeclaration := range importDeclarations {
+
+		// The import declaration imports from the same address
+		importedLocation, ok := importDeclaration.Location.(common.AddressLocation)
+		if !ok || importedLocation.Address != address {
+			continue
+		}
+
+		// The import declaration imports all identifiers, so also the location
+		if len(importDeclaration.Identifiers) == 0 {
+			return true
+		}
+
+		// The import declaration imports specific identifiers, so check if the location is imported
+		for _, identifier := range importDeclaration.Identifiers {
+			if identifier.Identifier == name {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func declaresConformanceTo(conformingDeclaration ast.ConformingDeclaration, name string) bool {
+	for _, conformance := range conformingDeclaration.ConformanceList() {
+		if conformance.Identifier.Identifier == name {
+			return true
+		}
+	}
+
+	return false
+}
+
+func isNominalType(ty ast.Type, name string) bool {
+	nominalType, ok := ty.(*ast.NominalType)
+	return ok &&
+		len(nominalType.NestedIdentifiers) == 0 &&
+		nominalType.Identifier.Identifier == name
+}
+
+const fungibleTokenTypeTotalSupplyFieldName = "totalSupply"
+const fungibleTokenVaultTypeBalanceFieldName = "balance"
+const fungibleTokenVaultTypeIdentifier = "Vault"
+
+func isFungibleTokenContract(program *ast.Program, fungibleTokenAddress common.Address) bool {
+	const contractName = "FungibleToken"
+
+	// Check if the contract imports the FungibleToken contract
+	if !importsAddressLocation(program, fungibleTokenAddress, contractName) {
+		return false
+	}
+
+	contractDeclaration := program.SoleContractDeclaration()
+	if contractDeclaration == nil {
+		return false
+	}
+
+	// Check if the contract implements the FungibleToken interface
+	if !declaresConformanceTo(contractDeclaration, contractName) {
+		return false
+	}
+
+	// Check if the contract has a totalSupply field
+	totalSupplyFieldDeclaration := getField(contractDeclaration, fungibleTokenTypeTotalSupplyFieldName)
+	if totalSupplyFieldDeclaration == nil {
+		return false
+	}
+
+	// Check if the totalSupply field is of type UFix64
+	if !isNominalType(totalSupplyFieldDeclaration.TypeAnnotation.Type, sema.UFix64TypeName) {
+		return false
+	}
+
+	// Check if the contract has a Vault resource
+
+	vaultDeclaration := contractDeclaration.Members.CompositesByIdentifier()[fungibleTokenVaultTypeIdentifier]
+	if vaultDeclaration == nil {
+		return false
+	}
+
+	// Check if the Vault resource has a balance field
+	balanceFieldDeclaration := getField(vaultDeclaration, fungibleTokenVaultTypeBalanceFieldName)
+	if balanceFieldDeclaration == nil {
+		return false
+	}
```

Comment:
Do we need to check these? Can we assume if a contract is deployed, then it must be a valid contract (and hence, given it conforms to the FungibleToken, it should have all these fields)?

Reply:
Yes, I guess we could maybe make these checks more lenient. Not 100% sure if that could introduce loopholes